# Data Manipulation tasks:
    
- Filtering data

- Sorting data

- Group By aggregations

- Adding new columns

- Merging dataframes

- Treating Missing Values

- Pivot Table

In [1]:
#import packages
import os
import pandas as pd
import numpy as np

In [7]:
#change the working directory
os.chdir("C://Users/shobhdas/Desktop\PYTHON-CODES-main\PYTHON-CODES-main/Week -2")

In [8]:
os.getcwd()

'C:\\Users\\shobhdas\\Desktop\\PYTHON-CODES-main\\PYTHON-CODES-main\\Week -2'

In [9]:
# load the data set 'stores.csv'
store=pd.read_csv("Store.csv")

In [10]:
#Number of observations in the data
store.shape

(9994, 22)

In [11]:
#Information about the columns/variables in a data set
store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     9994 non-null   int64  
 1   Row ID         9994 non-null   int64  
 2   Order ID       9994 non-null   object 
 3   Order Date     9994 non-null   object 
 4   Ship Date      9994 non-null   object 
 5   Ship Mode      9994 non-null   object 
 6   Customer ID    9994 non-null   object 
 7   Customer Name  9994 non-null   object 
 8   Segment        9994 non-null   object 
 9   Country        9994 non-null   object 
 10  City           9994 non-null   object 
 11  State          9994 non-null   object 
 12  Postal Code    9994 non-null   int64  
 13  Region         9994 non-null   object 
 14  Product ID     9994 non-null   object 
 15  Category       9994 non-null   object 
 16  Sub-Category   9994 non-null   object 
 17  Product Name   9994 non-null   object 
 18  Sales   

In [12]:
#First 3 observations from the dataset
store.head(3)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date     Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013  Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013  Second Class   
2           2       3  CA-2013-138688   6/13/2013   6/17/2013  Second Class   

  Customer ID    Customer Name    Segment        Country  ... Postal Code  \
0    CG-12520      Claire Gute   Consumer  United States  ...       42420   
1    CG-12520      Claire Gute   Consumer  United States  ...       42420   
2    DV-13045  Darrin Van Huff  Corporate  United States  ...       90036   

  Region       Product ID         Category Sub-Category  \
0  South  FUR-BO-10001798        Furniture    Bookcases   
1  South  FUR-CH-10000454        Furniture       Chairs   
2   West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.96        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.62        2   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820  
2       0.0    6.8714  

[3 rows x 22 columns]

In [13]:
#last 2 observations from the data set
store.tail(2)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date  \
9992        9992    9993  CA-2014-121258   2/27/2014  03-04-2014   
9993        9993    9994  CA-2014-119914  05-05-2014  05-10-2014   

           Ship Mode Customer ID Customer Name   Segment        Country  ...  \
9992  Standard Class    DB-13060   Dave Brooks  Consumer  United States  ...   
9993    Second Class    CC-12220  Chris Cortes  Consumer  United States  ...   

     Postal Code Region       Product ID         Category Sub-Category  \
9992       92627   West  OFF-PA-10004041  Office Supplies        Paper   
9993       92683   West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name   Sales Quantity  \
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.60        4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.16        2   

      Discount  Profit  
9992       0.0  13.320  
9993       0.0  72.948  

[2 rows x 22 columns]

In [14]:
#Count the number of missing values
store.isnull().sum()

Unnamed: 0       0
Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

In [15]:
#Summary statistics for numerical variables
store.describe()

Unnamed: 0       Row ID   Postal Code         Sales     Quantity  \
count  9994.000000  9994.000000   9994.000000   9994.000000  9994.000000   
mean   4996.500000  4997.500000  55190.379428    229.858001     3.789574   
std    2885.163629  2885.163629  32063.693350    623.245101     2.225110   
min       0.000000     1.000000   1040.000000      0.444000     1.000000   
25%    2498.250000  2499.250000  23223.000000     17.280000     2.000000   
50%    4996.500000  4997.500000  56430.500000     54.490000     3.000000   
75%    7494.750000  7495.750000  90008.000000    209.940000     5.000000   
max    9993.000000  9994.000000  99301.000000  22638.480000    14.000000   

          Discount       Profit  
count  9994.000000  9994.000000  
mean      0.156203    28.656896  
std       0.206452   234.260108  
min       0.000000 -6599.978000  
25%       0.000000     1.728750  
50%       0.200000     8.666500  
75%       0.200000    29.364000  
max       0.800000  8399.976000

In [ ]:
#mean>median (positivesqueu)
#mean=median(no sqeu)       


In [16]:
store.describe(include='all')

Unnamed: 0       Row ID        Order ID  Order Date   Ship Date  \
count   9994.000000  9994.000000            9994        9994        9994   
unique          NaN          NaN            5009        1238        1334   
top             NaN          NaN  CA-2014-100111  09-06-2013  12/16/2012   
freq            NaN          NaN              14          38          35   
mean    4996.500000  4997.500000             NaN         NaN         NaN   
std     2885.163629  2885.163629             NaN         NaN         NaN   
min        0.000000     1.000000             NaN         NaN         NaN   
25%     2498.250000  2499.250000             NaN         NaN         NaN   
50%     4996.500000  4997.500000             NaN         NaN         NaN   
75%     7494.750000  7495.750000             NaN         NaN         NaN   
max     9993.000000  9994.000000             NaN         NaN         NaN   

             Ship Mode Customer ID  Customer Name   Segment        Country  \
count             9994        9994           9994      9994           9994   
unique               4         793            793         3              1   
top     Standard Class    WB-21850  William Brown  Consumer  United States   
freq              5968          37             37      5191           9994   
mean               NaN         NaN            NaN       NaN            NaN   
std                NaN         NaN            NaN       NaN            NaN   
min                NaN         NaN            NaN       NaN            NaN   
25%                NaN         NaN            NaN       NaN            NaN   
50%                NaN         NaN            NaN       NaN            NaN   
75%                NaN         NaN            NaN       NaN            NaN   
max                NaN         NaN            NaN       NaN            NaN   

        ...   Postal Code Region       Product ID         Category  \
count   ...   9994.000000   9994             9994             9994   
unique  ...           NaN      4             1862                3   
top     ...           NaN   West  OFF-PA-10001970  Office Supplies   
freq    ...           NaN   3203               19             6026   
mean    ...  55190.379428    NaN              NaN              NaN   
std     ...  32063.693350    NaN              NaN              NaN   
min     ...   1040.000000    NaN              NaN              NaN   
25%     ...  23223.000000    NaN              NaN              NaN   
50%     ...  56430.500000    NaN              NaN              NaN   
75%     ...  90008.000000    NaN              NaN              NaN   
max     ...  99301.000000    NaN              NaN              NaN   

       Sub-Category     Product Name         Sales     Quantity     Discount  \
count          9994             9994   9994.000000  9994.000000  9994.000000   
unique           17             1850           NaN          NaN          NaN   
top         Binders  Staple envelope           NaN          NaN          NaN   
freq           1523               48           NaN          NaN          NaN   
mean            NaN              NaN    229.858001     3.789574     0.156203   
std             NaN              NaN    623.245101     2.225110     0.206452   
min             NaN              NaN      0.444000     1.000000     0.000000   
25%             NaN              NaN     17.280000     2.000000     0.000000   
50%             NaN              NaN     54.490000     3.000000     0.200000   
75%             NaN              NaN    209.940000     5.000000     0.200000   
max             NaN              NaN  22638.480000    14.000000     0.800000   

             Profit  
count   9994.000000  
unique          NaN  
top             NaN  
freq            NaN  
mean      28.656896  
std      234.260108  
min    -6599.978000  
25%        1.728750  
50%        8.666500  
75%       29.364000  
max     8399.976000  

[11 rows x 22 columns]

In [19]:
store.columns

Index(['Unnamed: 0', 'Row ID', 'Order ID', 'Order Date', 'Ship Date',
       'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country',
       'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='object')

In [18]:
# Finding out how many cities are the orders being delivered to
# We have to print the contents of the city column
cities=store['City'].unique()
len(cities)


531

In [20]:
#Unique levels in the Segment variable
store['Segment'].unique()

array(['Consumer', 'Corporate', 'Home Office'], dtype=object)

# Filter

##### Filter the data by segments are consumer

In [ ]:
#logical indexation [data[logical condition]]
#query method [data.query('logical condition')]

In [24]:
#logical indexation
store1=store[store['Segment']=='Consumer']
store1.head(5)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
3           3       4  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
4           4       5  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
5           5       6  CA-2011-115812  06-09-2011   6/14/2011  Standard Class   

  Customer ID    Customer Name   Segment        Country  ... Postal Code  \
0    CG-12520      Claire Gute  Consumer  United States  ...       42420   
1    CG-12520      Claire Gute  Consumer  United States  ...       42420   
3    SO-20335   Sean O'Donnell  Consumer  United States  ...       33311   
4    SO-20335   Sean O'Donnell  Consumer  United States  ...       33311   
5    BH-11710  Brosina Hoffman  Consumer  United States  ...       90032   

  Region       Product ID         Category Sub-Category  \
0  South  FUR-BO-10001798        Furniture    Bookcases   
1  South  FUR-CH-10000454        Furniture       Chairs   
3  South  FUR-TA-10000577        Furniture       Tables   
4  South  OFF-ST-10000760  Office Supplies      Storage   
5   West  FUR-FU-10001487        Furniture  Furnishings   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   
5  Eldon Expressions Wood and Plastic Desk Access...   48.8600        7   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  

[5 rows x 22 columns]

In [25]:
#query method
store.query('Segment=="Consumer"').head(5)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
3           3       4  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
4           4       5  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
5           5       6  CA-2011-115812  06-09-2011   6/14/2011  Standard Class   

  Customer ID    Customer Name   Segment        Country  ... Postal Code  \
0    CG-12520      Claire Gute  Consumer  United States  ...       42420   
1    CG-12520      Claire Gute  Consumer  United States  ...       42420   
3    SO-20335   Sean O'Donnell  Consumer  United States  ...       33311   
4    SO-20335   Sean O'Donnell  Consumer  United States  ...       33311   
5    BH-11710  Brosina Hoffman  Consumer  United States  ...       90032   

  Region       Product ID         Category Sub-Category  \
0  South  FUR-BO-10001798        Furniture    Bookcases   
1  South  FUR-CH-10000454        Furniture       Chairs   
3  South  FUR-TA-10000577        Furniture       Tables   
4  South  OFF-ST-10000760  Office Supplies      Storage   
5   West  FUR-FU-10001487        Furniture  Furnishings   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   
5  Eldon Expressions Wood and Plastic Desk Access...   48.8600        7   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  

[5 rows x 22 columns]

In [26]:
#Multiple conditions
#How many orders were placed for region south and for segment consumer
#data[(l1)&(l2)&(l3)]
#data[(l1)|(l2)]
store[(store['Region']=="South")& (store['Segment']=="Consumer")]

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date  \
0              0       1  CA-2013-152156  11-09-2013  11-12-2013   
1              1       2  CA-2013-152156  11-09-2013  11-12-2013   
3              3       4  US-2012-108966  10-11-2012  10/18/2012   
4              4       5  US-2012-108966  10-11-2012  10/18/2012   
12            12      13  CA-2014-114412   4/16/2014   4/21/2014   
...          ...     ...             ...         ...         ...   
9856        9856    9857  CA-2011-120950  11-06-2011  11-11-2011   
9881        9881    9882  CA-2011-153927  08-12-2011   8/13/2011   
9882        9882    9883  CA-2011-153927  08-12-2011   8/13/2011   
9980        9980    9981  US-2012-151435  09-06-2012  09-09-2012   
9989        9989    9990  CA-2011-110422   1/22/2011   1/24/2011   

           Ship Mode Customer ID       Customer Name   Segment        Country  \
0       Second Class    CG-12520         Claire Gute  Consumer  United States   
1       Second Class    CG-12520         Claire Gute  Consumer  United States   
3     Standard Class    SO-20335      Sean O'Donnell  Consumer  United States   
4     Standard Class    SO-20335      Sean O'Donnell  Consumer  United States   
12    Standard Class    AA-10480        Andrew Allen  Consumer  United States   
...              ...         ...                 ...       ...            ...   
9856  Standard Class    GA-14515     George Ashbrook  Consumer  United States   
9881     First Class    LL-16840  Lauren Leatherbury  Consumer  United States   
9882     First Class    LL-16840  Lauren Leatherbury  Consumer  United States   
9980    Second Class    SW-20455         Shaun Weien  Consumer  United States   
9989    Second Class    TB-21400    Tom Boeckenhauer  Consumer  United States   

      ... Postal Code Region       Product ID         Category Sub-Category  \
0     ...       42420  South  FUR-BO-10001798        Furniture    Bookcases   
1     ...       42420  South  FUR-CH-10000454        Furniture       Chairs   
3     ...       33311  South  FUR-TA-10000577        Furniture       Tables   
4     ...       33311  South  OFF-ST-10000760  Office Supplies      Storage   
12    ...       28027  South  OFF-PA-10002365  Office Supplies        Paper   
...   ...         ...    ...              ...              ...          ...   
9856  ...       31907  South  OFF-PA-10000587  Office Supplies        Paper   
9881  ...       30188  South  OFF-BI-10000138  Office Supplies      Binders   
9882  ...       30188  South  TEC-AC-10000023       Technology  Accessories   
9980  ...       70506  South  FUR-TA-10001039        Furniture       Tables   
9989  ...       33180  South  FUR-FU-10001889        Furniture  Furnishings   

                                           Product Name     Sales Quantity  \
0                     Bush Somerset Collection Bookcase  261.9600        2   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
3         Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                        Eldon Fold 'N Roll Cart System   22.3680        2   
12                                           Xerox 1967   15.5520        3   
...                                                 ...       ...      ...   
9856             Array Parchment Paper, Assorted Colors   43.6800        6   
9881                 Acco Translucent Poly Ring Binders   14.0400        3   
9882             Maxell 74 Minute CD-R Spindle, 50/Pack  272.6100       13   
9980                         KI Adjustable-Height Table   85.9800        1   
9989                             Ultra Door Pull Handle   25.2480        3   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
3         0.45 -383.0310  
4         0.20    2.5164  
12        0.20    5.4432  
...        ...       ...  
9856      0.00   20.9664  
9881      0.00    6.7392  
9882      0.00   98.1396  
9980      0.00   22.3548  
9989      0.20    4.1028  

[838 

In [27]:
store[(store['Region']=="South")& (store['Segment']=="Consumer")].shape[0]

838

838 orders were placed for region south and for segment consumer

In [30]:
#data.query('l1 & l2')
store.query('Region=="South" & Segment=="Consumer"').shape[0]

838

In [32]:
store.query('Profit>50.0').shape[0]

1728

In [ ]:
store1.to_csv("#file path")

# Sorting

In [33]:
#sort_values() method to sort the values
#sort data by profit
store.sort_values("Profit").head(2)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date  \
7772        7772    7773  CA-2013-108196  11/26/2013  12-03-2013   
683          683     684  US-2014-168116  11-05-2014  11-05-2014   

           Ship Mode Customer ID   Customer Name    Segment        Country  \
7772  Standard Class    CS-12505   Cindy Stewart   Consumer  United States   
683         Same Day    GT-14635  Grant Thornton  Corporate  United States   

      ... Postal Code Region       Product ID    Category Sub-Category  \
7772  ...       43130   East  TEC-MA-10000418  Technology     Machines   
683   ...       27217  South  TEC-MA-10004125  Technology     Machines   

                                   Product Name     Sales Quantity  Discount  \
7772  Cubify CubeX 3D Printer Double Head Print  4499.985        5       0.7   
683   Cubify CubeX 3D Printer Triple Head Print  7999.980        4       0.5   

         Profit  
7772 -6599.9780  
683  -3839.9904  

[2 rows x 22 columns]

In [34]:
#descending order
store.sort_values("Profit",ascending=False).head(2)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date  \
6826        6826    6827  CA-2013-118689  10-03-2013  10-10-2013   
8153        8153    8154  CA-2014-140151   3/24/2014   3/26/2014   

           Ship Mode Customer ID Customer Name    Segment        Country  ...  \
6826  Standard Class    TC-20980  Tamara Chand  Corporate  United States  ...   
8153     First Class    RB-19360  Raymond Buch   Consumer  United States  ...   

     Postal Code   Region       Product ID    Category Sub-Category  \
6826       47905  Central  TEC-CO-10004722  Technology      Copiers   
8153       98115     West  TEC-CO-10004722  Technology      Copiers   

                               Product Name     Sales Quantity  Discount  \
6826  Canon imageCLASS 2200 Advanced Copier  17499.95        5       0.0   
8153  Canon imageCLASS 2200 Advanced Copier  13999.96        4       0.0   

         Profit  
6826  8399.9760  
8153  6719.9808  

[2 rows x 22 columns]

In [35]:
# Which are the most valuable customers in South Region by Sales?
store.query('Region=="South"').sort_values("Sales",ascending=False).head(3)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date  \
2697        2697    2698  CA-2011-145317   3/18/2011   3/23/2011   
8488        8488    8489  CA-2013-158841  02-02-2013  02-04-2013   
683          683     684  US-2014-168116  11-05-2014  11-05-2014   

           Ship Mode Customer ID   Customer Name      Segment        Country  \
2697  Standard Class    SM-20320     Sean Miller  Home Office  United States   
8488    Second Class    SE-20110    Sanjit Engle     Consumer  United States   
683         Same Day    GT-14635  Grant Thornton    Corporate  United States   

      ... Postal Code Region       Product ID    Category Sub-Category  \
2697  ...       32216  South  TEC-MA-10002412  Technology     Machines   
8488  ...       22204  South  TEC-MA-10001127  Technology     Machines   
683   ...       27217  South  TEC-MA-10004125  Technology     Machines   

                                           Product Name     Sales Quantity  \
2697  Cisco TelePresence System EX90 Videoconferenci...  22638.48        6   
8488  HP Designjet T520 Inkjet Large Format Printer ...   8749.95        5   
683           Cubify CubeX 3D Printer Triple Head Print   7999.98        4   

      Discount     Profit  
2697       0.5 -1811.0784  
8488       0.0  2799.9840  
683        0.5 -3839.9904  

[3 rows x 22 columns]

In [36]:
#sort by two variables
store.sort_values(["Profit","Sales"],ascending=[False,True]).head(2)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date  \
6826        6826    6827  CA-2013-118689  10-03-2013  10-10-2013   
8153        8153    8154  CA-2014-140151   3/24/2014   3/26/2014   

           Ship Mode Customer ID Customer Name    Segment        Country  ...  \
6826  Standard Class    TC-20980  Tamara Chand  Corporate  United States  ...   
8153     First Class    RB-19360  Raymond Buch   Consumer  United States  ...   

     Postal Code   Region       Product ID    Category Sub-Category  \
6826       47905  Central  TEC-CO-10004722  Technology      Copiers   
8153       98115     West  TEC-CO-10004722  Technology      Copiers   

                               Product Name     Sales Quantity  Discount  \
6826  Canon imageCLASS 2200 Advanced Copier  17499.95        5       0.0   
8153  Canon imageCLASS 2200 Advanced Copier  13999.96        4       0.0   

         Profit  
6826  8399.9760  
8153  6719.9808  

[2 rows x 22 columns]

# Group By
# Groupby task(split the data into groups based on some criteria)

In [41]:
# I need the sum of the sale amount for each state 

state_group=store.groupby('State', as_index=False)
#store.groupby('State').reset_index()
state_group

#agg method

In [42]:
sales_sum=state_group.agg({"Sales":'sum'})

In [43]:
sales_sum.head(2)

State      Sales
0  Alabama  19510.640
1  Arizona  35282.001

In [44]:
#Filter the data by East Region, find the maximum sales by State wise
store.query('Region=="East"').groupby('State').agg({"Sales":np.max}).reset_index().head()

State     Sales
0           Connecticut   1133.35
1              Delaware  10499.97
2  District of Columbia   1379.92
3                 Maine    437.85
4              Maryland   2541.98

In [49]:
# What is the average quantity sold by region?
store.groupby('Region').agg({"Quantity":np.mean}).reset_index()

Region  Quantity
0  Central  3.779595
1     East  3.728230
2    South  3.832716
3     West  3.829535

In [52]:
#Find out the top 5 states in terms of sale amount 
store.groupby('State',as_index=False).agg({'Sales':'sum'}).sort_values("Sales",ascending=False).head()

State        Sales
3     California  457687.6315
30      New York  310876.2710
41         Texas  170188.0458
45    Washington  138641.2700
36  Pennsylvania  116511.9140

# Renaming the columns

In [57]:
store3=store.rename(columns={'Product Name':'Product_Name'})
store3.head()

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
2           2       3  CA-2013-138688   6/13/2013   6/17/2013    Second Class   
3           3       4  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
4           4       5  US-2012-108966  10-11-2012  10/18/2012  Standard Class   

  Customer ID    Customer Name    Segment        Country  ... Postal Code  \
0    CG-12520      Claire Gute   Consumer  United States  ...       42420   
1    CG-12520      Claire Gute   Consumer  United States  ...       42420   
2    DV-13045  Darrin Van Huff  Corporate  United States  ...       90036   
3    SO-20335   Sean O'Donnell   Consumer  United States  ...       33311   
4    SO-20335   Sean O'Donnell   Consumer  United States  ...       33311   

  Region       Product ID         Category Sub-Category  \
0  South  FUR-BO-10001798        Furniture    Bookcases   
1  South  FUR-CH-10000454        Furniture       Chairs   
2   West  OFF-LA-10000240  Office Supplies       Labels   
3  South  FUR-TA-10000577        Furniture       Tables   
4  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product_Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 22 columns]

In [56]:
#Rename Sales as "Average Sales"
store2=store.rename(columns={'Sales':'Average sales'})
store2.head(2)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date     Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013  Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013  Second Class   

  Customer ID Customer Name   Segment        Country  ... Postal Code Region  \
0    CG-12520   Claire Gute  Consumer  United States  ...       42420  South   
1    CG-12520   Claire Gute  Consumer  United States  ...       42420  South   

        Product ID   Category Sub-Category  \
0  FUR-BO-10001798  Furniture    Bookcases   
1  FUR-CH-10000454  Furniture       Chairs   

                                        Product Name Average sales Quantity  \
0                  Bush Somerset Collection Bookcase        261.96        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...        731.94        3   

   Discount    Profit  
0       0.0   41.9136  
1       0.0  219.5820  

[2 rows x 22 columns]

# Adding new columns

In [58]:
## Adding new columns
# Comparing Sales per order with the average sales 
store["Hi_low"]=(store['Sales'])/(store['Sales'].mean())

In [59]:
store.head(2)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date     Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013  Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013  Second Class   

  Customer ID Customer Name   Segment        Country  ... Region  \
0    CG-12520   Claire Gute  Consumer  United States  ...  South   
1    CG-12520   Claire Gute  Consumer  United States  ...  South   

        Product ID   Category Sub-Category  \
0  FUR-BO-10001798  Furniture    Bookcases   
1  FUR-CH-10000454  Furniture       Chairs   

                                        Product Name   Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.96        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94        3   

  Discount    Profit    Hi_low  
0      0.0   41.9136  1.139660  
1      0.0  219.5820  3.184314  

[2 rows x 23 columns]

In [66]:
#Create Cost price variable
store["Cost_Price"]=(store['Sales']-store['Profit'])/store['Quantity']
store.head()

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
2           2       3  CA-2013-138688   6/13/2013   6/17/2013    Second Class   
3           3       4  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
4           4       5  US-2012-108966  10-11-2012  10/18/2012  Standard Class   

  Customer ID    Customer Name    Segment        Country  ...  \
0    CG-12520      Claire Gute   Consumer  United States  ...   
1    CG-12520      Claire Gute   Consumer  United States  ...   
2    DV-13045  Darrin Van Huff  Corporate  United States  ...   
3    SO-20335   Sean O'Donnell   Consumer  United States  ...   
4    SO-20335   Sean O'Donnell   Consumer  United States  ...   

        Product ID         Category  Sub-Category  \
0  FUR-BO-10001798        Furniture     Bookcases   
1  FUR-CH-10000454        Furniture        Chairs   
2  OFF-LA-10000240  Office Supplies        Labels   
3  FUR-TA-10000577        Furniture        Tables   
4  OFF-ST-10000760  Office Supplies       Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit    Hi_low  Cost_Price  
0     0.00   41.9136  1.139660    110.0232  
1     0.00  219.5820  3.184314    170.7860  
2     0.00    6.8714  0.063604      3.8743  
3     0.45 -383.0310  4.165952    268.1217  
4     0.20    2.5164  0.097312      9.9258  

[5 rows x 24 columns]

# Apply method in a data frame

In [67]:
#Apply method
# What is the Total Sales, Quantity, Discount, Profit .

def get_sum(x):
    return np.sum(x)


store[["Sales","Quantity","Discount","Profit"]].apply(get_sum,axis=0)
#axis=0---> along the rows
#axis=1---> along the col

Sales       2.297201e+06
Quantity    3.787300e+04
Discount    1.561090e+03
Profit      2.863970e+05
dtype: float64

In [68]:
#region should  be south and ship mode will be second class
def new1(x):
    if(x["Region"]=='South')| (x["Ship Mode"]=='Second Class'):
        return 'high'
    else:
        return 'low'
store['new1']=store.apply(new1,axis=1)

In [69]:
store.head(1)

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date     Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013  Second Class   

  Customer ID Customer Name   Segment        Country  ...   Category  \
0    CG-12520   Claire Gute  Consumer  United States  ...  Furniture   

  Sub-Category                       Product Name   Sales Quantity Discount  \
0    Bookcases  Bush Somerset Collection Bookcase  261.96        2      0.0   

    Profit   Hi_low  Cost_Price  new1  
0  41.9136  1.13966    110.0232  high  

[1 rows x 25 columns]

# map function

In [70]:
#create new variable 'salesnew'from 'Sales' variable (condition: if sales>3000,return 'high',else return 'low')
def flag(x):
    if x>3000:
        return 'high'
    else:
        return 'low'
store['salesnew']=store['Sales'].map(flag)
store.head()

Unnamed: 0  Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0           0       1  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
1           1       2  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
2           2       3  CA-2013-138688   6/13/2013   6/17/2013    Second Class   
3           3       4  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
4           4       5  US-2012-108966  10-11-2012  10/18/2012  Standard Class   

  Customer ID    Customer Name    Segment        Country  ... Sub-Category  \
0    CG-12520      Claire Gute   Consumer  United States  ...    Bookcases   
1    CG-12520      Claire Gute   Consumer  United States  ...       Chairs   
2    DV-13045  Darrin Van Huff  Corporate  United States  ...       Labels   
3    SO-20335   Sean O'Donnell   Consumer  United States  ...       Tables   
4    SO-20335   Sean O'Donnell   Consumer  United States  ...      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

  Discount    Profit    Hi_low Cost_Price  new1  salesnew  
0     0.00   41.9136  1.139660   110.0232  high       low  
1     0.00  219.5820  3.184314   170.7860  high       low  
2     0.00    6.8714  0.063604     3.8743  high       low  
3     0.45 -383.0310  4.165952   268.1217  high       low  
4     0.20    2.5164  0.097312     9.9258  high       low  

[5 rows x 26 columns]

# Combining data frames using concat function

In [73]:

data1=pd.read_csv("Sales_Sep.csv")
data2=pd.read_csv("Sales_Oct.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'Sales_Sep.csv'

In [ ]:
data1.head()

In [ ]:
data2.head()

In [ ]:
# Concatenate 
row__concat

In [ ]:
#export a file to working directory


# Combining data frames using merge function 

In [74]:
## Merging DataFrames
df1=pd.DataFrame({'CustomerID':[1,2,3,4,5,6],'Product':['Television','Television','Television',
                                                        'Earphones','Earphones','Earphones']})
df2=pd.DataFrame({'CustomerID':[2,4,6],'State':['Texas','Texas','Seattle']})

In [75]:
#outer join:Every records will appear
#pd.merge(df1,df2,how='outer', on='CustomerID')
pd.merge(df1,df2,how='outer', on='CustomerID')


CustomerID     Product    State
0           1  Television      NaN
1           2  Television    Texas
2           3  Television      NaN
3           4   Earphones    Texas
4           5   Earphones      NaN
5           6   Earphones  Seattle

In [76]:
#inner join: only matcing records will appear
pd.merge(df1,df2,how='inner', on='CustomerID')


CustomerID     Product    State
0           2  Television    Texas
1           4   Earphones    Texas
2           6   Earphones  Seattle

In [78]:

#left merge:
#Data frame df1 will be keep as it as and on the basis of common variable merge dataframe df2
pd.merge(df1,df2,how='left', on='CustomerID')


CustomerID     Product    State
0           1  Television      NaN
1           2  Television    Texas
2           3  Television      NaN
3           4   Earphones    Texas
4           5   Earphones      NaN
5           6   Earphones  Seattle

In [79]:
#right join: df2 will be keep as it is and on the basis of common variable merge data frame df1
pd.merge(df1,df2,how='right', on='CustomerID')


CustomerID     Product    State
0           2  Television    Texas
1           4   Earphones    Texas
2           6   Earphones  Seattle

In [91]:
#drop a column
df1.drop('CustomerID',axis=1)


Product
0  Television
1  Television
2  Television
3   Earphones
4   Earphones
5   Earphones

In [96]:
store.drop('Unnamed: 0',axis=1)


Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
0          1  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
1          2  CA-2013-152156  11-09-2013  11-12-2013    Second Class   
2          3  CA-2013-138688   6/13/2013   6/17/2013    Second Class   
3          4  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
4          5  US-2012-108966  10-11-2012  10/18/2012  Standard Class   
...      ...             ...         ...         ...             ...   
9989    9990  CA-2011-110422   1/22/2011   1/24/2011    Second Class   
9990    9991  CA-2014-121258   2/27/2014  03-04-2014  Standard Class   
9991    9992  CA-2014-121258   2/27/2014  03-04-2014  Standard Class   
9992    9993  CA-2014-121258   2/27/2014  03-04-2014  Standard Class   
9993    9994  CA-2014-119914  05-05-2014  05-10-2014    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United States        Henderson   
1       CG-12520       Claire Gute   Consumer  United States        Henderson   
2       DV-13045   Darrin Van Huff  Corporate  United States      Los Angeles   
3       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
4       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...          ...               ...        ...            ...              ...   
9989    TB-21400  Tom Boeckenhauer   Consumer  United States            Miami   
9990    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9991    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9992    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9993    CC-12220      Chris Cortes   Consumer  United States      Westminster   

      ... Sub-Category                                       Product Name  \
0     ...    Bookcases                  Bush Somerset Collection Bookcase   
1     ...       Chairs  Hon Deluxe Fabric Upholstered Stacking Chairs,...   
2     ...       Labels  Self-Adhesive Address Labels for Typewriters b...   
3     ...       Tables      Bretford CR4500 Series Slim Rectangular Table   
4     ...      Storage                     Eldon Fold 'N Roll Cart System   
...   ...          ...                                                ...   
9989  ...  Furnishings                             Ultra Door Pull Handle   
9990  ...  Furnishings  Tenex B1-RE Series Chair Mats for Low Pile Car...   
9991  ...       Phones                              Aastra 57i VoIP phone   
9992  ...        Paper  It's Hot Message Books with Stickers, 2 3/4" x 5"   
9993  ...   Appliances  Acco 7-Outlet Masterpiece Power Center, Wihtou...   

         Sales Quantity Discount    Profit    Hi_low  Cost_Price  new1  \
0     261.9600        2     0.00   41.9136  1.139660    110.0232  high   
1     731.9400        3     0.00  219.5820  3.184314    170.7860  high   
2      14.6200        2     0.00    6.8714  0.063604      3.8743  high   
3     957.5775        5     0.45 -383.0310  4.165952    268.1217  high   
4      22.3680        2     0.20    2.5164  0.097312      9.9258  high   
...        ...      ...      ...       ...       ...         ...   ...   
9989   25.2480        3     0.20    4.1028  0.109842      7.0484  high   
9990   91.9600        2     0.00   15.6332  0.400073     38.1634   low   
9991  258.5760        2     0.20   19.3932  1.124938    119.5914   low   
9992   29.6000        4     0.00   13.3200  0.128775      4.0700   low   
9993  243.1600        2     0.00   72.9480  1.057871     85.1060  high   

      salesnew  
0          low  
1          low  
2          low  
3          low  
4          low  
...        ...  
9989       low  
9990       low  
9991       low  
9992       low  
9993       low  

[9994 rows x 25 columns]

In [85]:
store.columns

Index(['Unnamed: 0', 'Row ID', 'Order ID', 'Order Date', 'Ship Date',
       'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country',
       'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit', 'Hi_low', 'Cost_Price', 'new1', 'salesnew'],
      dtype='object')

# Handling Character data using dummies

In [93]:
dat=pd.read_csv("medal.csv",encoding='latin')
dat.head(2)

City  Edition     Sport Discipline           Athlete  NOC Gender  \
0  Athens     1896  Aquatics   Swimming     HAJOS, Alfred  HUN    Men   
1  Athens     1896  Aquatics   Swimming  HERSCHMANN, Otto  AUT    Men   

            Event Event_gender   Medal  
0  100m freestyle            M    Gold  
1  100m freestyle            M  Silver

In [94]:
dummies=pd.get_dummies(dat)
dummies.head(2)

Edition  City_Amsterdam  City_Antwerp  City_Athens  City_Atlanta  \
0     1896               0             0            1             0   
1     1896               0             0            1             0   

   City_Barcelona  City_Beijing  City_Berlin  City_Helsinki  City_London  ...  \
0               0             0            0              0            0  ...   
1               0             0            0              0            0  ...   

   Event_épée individual  Event_épée team  \
0                        0                  0   
1                        0                  0   

   Event_épée, amateurs and masters  Event_épée, masters  Event_gender_M  \
0                                   0                      0               1   
1                                   0                      0               1   

   Event_gender_W  Event_gender_X  Medal_Bronze  Medal_Gold  Medal_Silver  
0               0               0             0           1             0  
1               0               0             0           0             1  

[2 rows x 22257 columns]

# Handling missing values 

In [98]:
# Counting the number of missing values in each column
data_C=pd.read_csv("credit_history.csv",na_values=['Missing'])

# If there are any blank spaces or keyword Missing in the 
#source CSV file, they will be replaced with NAN

In [99]:
data_C.isnull().sum()

default        0
amount         0
grade          0
years        279
ownership      0
income         0
age            0
dtype: int64

In [100]:
data_C.isnull().sum()/data_C.shape[0]

default      0.000000
amount       0.000000
grade        0.000000
years        0.036107
ownership    0.000000
income       0.000000
age          0.000000
dtype: float64

In [102]:
#summary statistics for 'Year' variablE
data_C['years'].describe()

count    7727.000000
mean        6.011000
std         6.590163
min         0.000000
25%         2.000000
50%         4.000000
75%         8.000000
max        62.000000
Name: years, dtype: float64

In [103]:
#Replacing missing values
data_C['years']=data_C['years'].fillna(data_C['years'].median())

In [105]:
data_C.isnull().sum()

default      0
amount       0
grade        0
years        0
ownership    0
income       0
age          0
dtype: int64

In [ ]:
#<5%: remove only the missing observations
#5%<x<30%: dO IMPUTATION
             #continuous variables : mean/median (skewed data:median)
             #categorical variables :mode
#>30% : Remove the variable



### Pivot table
pandas.pivot(index, columns, values) function produces pivot table based on 3 columns of the DataFrame.

Uses unique values from index / columns and fills with values.

Python Pandas function pivot_table help us with the summarization and conversion of dataframe in long form to dataframe in wide form, in a variety of complex scenarios.

In [104]:
df = pd.DataFrame({'First Name': ['Aryan', 'Rohan', 'Riya', 'Yash', 'Siddhant', ],
                   'Last Name': ['Singh', 'Agarwal', 'Shah', 'Bhatia', 'Khanna'],
                   'Type': ['Full-time Employee', 'Intern', 'Full-time Employee', 
                            'Part-time Employee', 'Full-time Employee'],
                   'Department': ['Administration', 'Technical', 'Administration', 
                                  'Technical', 'Management'],
                   'YoE': [2, 3, 5, 7, 6],
                   'Salary': [20000, 5000, 10000, 10000, 20000]})

df

First Name Last Name                Type      Department  YoE  Salary
0      Aryan     Singh  Full-time Employee  Administration    2   20000
1      Rohan   Agarwal              Intern       Technical    3    5000
2       Riya      Shah  Full-time Employee  Administration    5   10000
3       Yash    Bhatia  Part-time Employee       Technical    7   10000
4   Siddhant    Khanna  Full-time Employee      Management    6   20000

Use pd.pivot_table and specify the data, index, columns, aggfunc and `values` parameters.

In [111]:
# create pivot table of average salary for employees based on their Department and Job type
out=pd.pivot_table(data=df,index=['Type'],columns=['Department'],values='Salary',aggfunc=['mean'])
out

mean                     
Department         Administration Management Technical
Type                                                  
Full-time Employee        15000.0    20000.0       NaN
Intern                        NaN        NaN    5000.0
Part-time Employee            NaN        NaN   10000.0

If column parameter is not specified, it will aggregate based on the index.

In [112]:
# Pivot table with multiple aggfuncs
# aggfunc=['sum','mean']
out=pd.pivot_table(data=df,index=['Type'],columns=['Department'],values='Salary',aggfunc=['mean','sum'])
out

mean                                 sum  \
Department         Administration Management Technical Administration   
Type                                                                    
Full-time Employee        15000.0    20000.0       NaN        30000.0   
Intern                        NaN        NaN    5000.0            NaN   
Part-time Employee            NaN        NaN   10000.0            NaN   

                                         
Department         Management Technical  
Type                                     
Full-time Employee    20000.0       NaN  
Intern                    NaN    5000.0  
Part-time Employee        NaN   10000.0

More use cases of the pivot_table function could be used

https://pbpython.com/pandas-pivot-table-explained.html

